File: Graphs/AdjacencyList/CourseSchedule.js
//Time: O(n) / O(N + P) number of nodes(courses) + preReqs 
//Space: O(n)

/**
 * @param {number} numCourses
 * @param {number[][]} prerequisites
 * @return {boolean}
 */
var canFinish = function (numCourses, prerequisites) {
    let preMap = {}

    for (let i = 0; i < numCourses; i++) {
        preMap[i] = []
    }

    console.log(preMap)

    for (const [crs, pre] of prerequisites) {
        preMap[crs].push(pre)
    }

    console.log(preMap)

    let visitSet = new Set

    const dfs = (crs) => {
        if (visitSet.has(crs)) {
            return false
        }
        if (preMap[crs] == []) {
            return true
        }

        visitSet.add(crs)

        for (const pre of preMap[crs]) {
            if (!dfs(pre)) {
                return false
            }
        }
        
        visitSet.delete(crs) //finished visiting and can take course
        preMap[crs] = [] //if we visit this node again we know we can take prereq so set it to empty array

        return true //course can be taken 
    }


    //have to manually iterate through courses like this incase we get a graph that is not entirely connected
    // 1 -> 2, 1 has preReq of 2 
    // 3 -> 4, 3 has preReq of 4
    //iterating like this fixes this issue
    for (let i = 0; i < numCourses; i++) {
        if (!dfs(i)) {
            return false
        }
    }

    return true
};

File: AdvancedAlgorithms/Graphs/TopologicalSort/CourseSchedule.py
#Where n is the number of courses/nodes and p is the prerequisites
#Time: O(n + p) 
#Space: O(n)

class Solution:
    def canFinish(self, numCourses: int, prerequisites: List[List[int]]) -> bool:
        preMap = {i: [] for i in range(numCourses)}

        for crs, pre in prerequisites:
            preMap[crs].append(pre)

        visit = set()

        def dfs(crs):
            if crs in visit:
                return False

            #no preReqs or preReqs left
            if preMap[crs] == []:
                return True

            visit.add(crs)

            for preReq in preMap[crs]:
                if not dfs(preReq): return False

                #The .remove(crs) was so confusing but I finally understood it. Simplest explanation: if we exit the for-loop inside dfs, we know that crs is a good node without cycles. 
                #However, if it remained in the visited set, we could trip the first if-clause in the dfs function if we revisit it and return False. 
                #That's what we don't want to do, because we just calculated that crs has no cycles. So we remove it from visited so that other paths can successfully revisit it.
                #Basically we can visit the node twice without it being a cycle due to it terminating multiple paths.
            visit.remove(crs)
            preMap[crs] = []
            return True

        for crs in range(numCourses):
            if not dfs(crs): return False

        return True